Closes #531: Initial frontend work for GatsbyJS app

miggs125 opened the pull request:
addresses #531.
<!--
1. Automatically close the issue when this PR is merged
    USAGE: Fixes #<issue number>
2. If your PR addresses an issue but does not close it
    USAGE: #<issue number> <reason>(i.e. This issue was worked on by @user and myself and his PR should close the issue.)
-->

## Type of Change

<!-- bug fix, feature, documentation, UI, etc. -->

- [ ] **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [x] **UI**: Change which improves UI

## Description

<!-- Please add a detailed description of what this PR does and why it is needed -->

## Checklist

<!-- Before submitting a PR, address each item -->

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

miggs125 commented:
> 
> 
> Do you think we need these changes in this PR?
> 
> ![Screenshot_20200116_204205](https://user-images.githubusercontent.com/23108901/72577341-6ba32c00-38a0-11ea-8fcf-94c4a59c87a0.png)

I added this by accident as I branched off of the ssl branch. Reverted in my last commit

cindyledev commented:
Oh it looks like you didn't run prettier and our tests got sad

agarcia-caicedo commented:
I think it looks pretty good, and there isn't any issues with the interface that I can see

miggs125 commented:
Not entirely sure why it's including all the commits from previous contributors. Anyway, I updated the work I had done and removed material UI, added a sideDrawer. Added some custom components. I think this should be a good base for those who are interested to begin contributing to the front end.

cindyledev commented:
If this lands, I found a bug for someone to fix in the future :P 
Not entirely sure why but the hamburger menu button doesn't seem to show up properly on my machine.
![VirtualBox_Fedora_20_01_2020_18_34_35](https://user-images.githubusercontent.com/36822198/72763906-ce076f80-3bb3-11ea-83c1-f3bacd95fcc1.png)
I can get to the side drawer only if I click on the exact pixel.

cindyledev commented:
Also I can scroll a little bit left and right which shouldn't be happening. Maybe the navbar is a little big wider than it should be? This needs further investigation

cindyledev commented:
My random nits can be fixed in future issues. I'd like this to pass our tests ASAP since this PR might be blocking whatever @agarcia-caicedo is doing in the front end. Pretty sure it's just linting and prettier errors.

humphd commented on .gitignore:
```diff
@@ -40,3 +40,9 @@ dump.rdb
 certs/
+
+# gastby ,cache directory
+.cache/
```
These folders will be under `src/frontend` right?  We should prefix

miggs125 replied:
yes these are under frontend. By prefix do you mean writing out the full path: `./src/frontend/.cache/` ?

humphd replied:
Right, exactly.  Otherwise it's going to ignore them in the root, where they don't exist.

humphd commented on src/frontend/gatsby-config.js:
```diff
@@ -0,0 +1,34 @@
+module.exports = {
+  siteMetadata: {
+    title: `Gatsby Default Starter`,
```
Let's update this to be own the code.

humphd commented on src/frontend/package.json:
```diff
@@ -0,0 +1,51 @@
+{
+  "name": "gatsby-starter-default",
+  "private": true,
+  "description": "A simple starter to get up and developing quickly with Gatsby",
```
Update this, author, etc.

humphd commented on src/frontend/package.json:
```diff
@@ -0,0 +1,51 @@
+  "devDependencies": {
+    "json-loader": "^0.5.7",
+    "node-sass": "^4.13.1",
+    "prettier": "^1.19.1",
```
Remove this, we'll do it from the root

humphd commented on src/frontend/package.json:
```diff
@@ -0,0 +1,51 @@
+  "scripts": {
+    "build": "gatsby build",
+    "develop": "gatsby develop",
+    "format": "prettier --write \"**/*.{js,jsx,json,md}\"",
```
Remove this, we'll manage Prettier at the root.

humphd commented on src/frontend/package.json:
```diff
@@ -0,0 +1,51 @@
+    "build": "gatsby build",
+    "develop": "gatsby develop",
+    "format": "prettier --write \"**/*.{js,jsx,json,md}\"",
+    "start": "npm run develop",
```
I wonder if we should lift some of these up to the root?  We could add `frontend-*` scripts to do build, develop and clean.  That might be all we really need, and we can do:

```
npm run frontend-build
npm run frontend-develop
etc
```

miggs125 replied:
Good idea. I will add these scripts to package.json

humphd commented on src/frontend/package.json:
```diff
@@ -0,0 +1,51 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/gatsbyjs/gatsby-starter-default"
```
Remove this whole section.

humphd commented on src/frontend/package.json:
```diff
@@ -0,0 +1,51 @@
+    "url": "https://github.com/gatsbyjs/gatsby-starter-default"
+  },
+  "bugs": {
+    "url": "https://github.com/gatsbyjs/gatsby/issues"
```
Same with this.

humphd commented on src/frontend/src/components/layout.js:
```diff
@@ -0,0 +1,72 @@
+
+import logo from '../images/logo.svg';
+
+import data from '../../../../dummyData.json';
```
This looks like it was deleted above, should we update this?

humphd commented on src/frontend/src/components/layout.js:
```diff
@@ -0,0 +1,72 @@
+    });
+  }
+
+  drawerToggle = () => {
```
This is failing linting because it's a proposal not in ecmascript yet.  You either need to add babel and deal with this transform, or (better) just use standard ES6 classes and `.bind()` in the constructor so you can rely on `this` here.

miggs125 replied:
> 
> 
> If this lands, I found a bug for someone to fix in the future :P
> Not entirely sure why but the hamburger menu button doesn't seem to show up properly on my machine.
> ![VirtualBox_Fedora_20_01_2020_18_34_35](https://user-images.githubusercontent.com/36822198/72763906-ce076f80-3bb3-11ea-83c1-f3bacd95fcc1.png)
> I can get to the side drawer only if I click on the exact pixel.

I did not finish the hamburger button, up for grabs.

humphd commented on src/frontend/src/sass/abstracts/_variables.scss:
```diff
@@ -0,0 +1,12 @@
+// colors
+$color-primary-dark: #375e77;
```
If this is all we're using SASS for, I'd say it's overkill.  Let's just use CSS variables and dump the extra complexity.

miggs125 replied:
I primarily added sass because it allows scoping and variable substitution which makes future style changes easier to implement. I agree though that if the complexity is not worth the returns then we can get rid of it

humphd replied:
I leave it up to those of you who will live in the CSS.  If we need it, keep it.  If we don't, it's one more thing people have to learn to get involved.

miggs125 replied:
@agarcia-caicedo what do you think? Would you like to keep the sass infrastructure?

agarcia-caicedo replied:
Like David said, it is an added level of complexity, and while I can see value from using sass, I don't think this project sufficiently benefits from its use

miggs125 replied:
I'll get rid of it then!

humphd commented:
@miggs125 @agarcia-caicedo Can we also please add a static route in our Express server so we can serve the built site?  Should that happen in this PR or do I need to file another one?

agarcia-caicedo commented:
> 
> 
> @miggs125 @agarcia-caicedo Can we also please add a static route in our Express server so we can serve the built site? Should that happen in this PR or do I need to file another one?

@humpdh, I think that should go on a different issue

humphd commented:
@agarcia-caicedo ok sounds good, filed #569.

miggs125 commented:
> 
> 
> If this lands, I found a bug for someone to fix in the future :P
> Not entirely sure why but the hamburger menu button doesn't seem to show up properly on my machine.
> ![VirtualBox_Fedora_20_01_2020_18_34_35](https://user-images.githubusercontent.com/36822198/72763906-ce076f80-3bb3-11ea-83c1-f3bacd95fcc1.png)
> I can get to the side drawer only if I click on the exact pixel.

I did not finish the hamburger button, up for grabs

humphd commented:
@miggs125 thanks for filing those follow-ups.  For next time, prefer deep linking into a long discussion (e.g,. https://github.com/Seneca-CDOT/telescope/pull/541#discussion_r369188779) vs. to the top-level of a PR (e.g., https://github.com/Seneca-CDOT/telescope/pull/541) when splitting off work.  This helps people find what you mean more quickly.

humphd commented:
Can we get an update on this?  Is it landing, needing changes, waiting on reviews?  Let's prioritize it please.
